File: examples/javascript/src/vector3.js
/** Immutable 3-D vector with operator-style methods. */
export class Vector3 {
  #x; #y; #z;

  constructor(x = 0, y = 0, z = 0) {
    this.#x = x;
    this.#y = y;
    this.#z = z;
    Object.freeze(this);
  }

  get x() { return this.#x; }
  get y() { return this.#y; }
  get z() { return this.#z; }

  length() {
    return Math.sqrt(this.#x ** 2 + this.#y ** 2 + this.#z ** 2);
  }

  normalized() {
    const len = this.length();
    return len === 0 ? new Vector3() : new Vector3(this.#x / len, this.#y / len, this.#z / len);
  }

  add(other) { return new Vector3(this.#x + other.x, this.#y + other.y, this.#z + other.z); }
  sub(other) { return new Vector3(this.#x - other.x, this.#y - other.y, this.#z - other.z); }
  scale(s)   { return new Vector3(this.#x * s, this.#y * s, this.#z * s); }

  dot(other) {
    return this.#x * other.x + this.#y * other.y + this.#z * other.z;
  }

  cross(other) {
    return new Vector3(
      this.#y * other.z - this.#z * other.y,
      this.#z * other.x - this.#x * other.z,
      this.#x * other.y - this.#y * other.x,
    );
  }

  // Symbol.iterator — enables destructuring: const [x, y, z] = vec
  [Symbol.iterator]() {
    return [this.#x, this.#y, this.#z][Symbol.iterator]();
  }

  toString() {
    return `(${this.#x}, ${this.#y}, ${this.#z})`;
  }
}

File: examples/python/src/vector3.py
"""3D vector with operator overloads and generic type support."""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Generic, Iterator, TypeVar

T = TypeVar("T", int, float)


@dataclass(frozen=True, slots=True)
class Vector3(Generic[T]):
    x: T
    y: T
    z: T

    def length(self) -> float:
        return math.sqrt(self.x**2 + self.y**2 + self.z**2)

    def normalized(self) -> Vector3[float]:
        ln = self.length()
        if ln == 0:
            return Vector3(0.0, 0.0, 0.0)
        return Vector3(self.x / ln, self.y / ln, self.z / ln)

    def dot(self, other: Vector3[T]) -> T:
        return self.x * other.x + self.y * other.y + self.z * other.z

    def cross(self, other: Vector3[T]) -> Vector3[T]:
        return Vector3(
            self.y * other.z - self.z * other.y,
            self.z * other.x - self.x * other.z,
            self.x * other.y - self.y * other.x,
        )

    def __add__(self, other: Vector3[T]) -> Vector3[T]:
        return Vector3(self.x + other.x, self.y + other.y, self.z + other.z)

    def __sub__(self, other: Vector3[T]) -> Vector3[T]:
        return Vector3(self.x - other.x, self.y - other.y, self.z - other.z)

    def __mul__(self, scalar: T) -> Vector3[T]:
        return Vector3(self.x * scalar, self.y * scalar, self.z * scalar)

    def __iter__(self) -> Iterator[T]:
        yield self.x
        yield self.y
        yield self.z

    def __repr__(self) -> str:
        return f"({self.x}, {self.y}, {self.z})"


Vec3f = Vector3[float]
Vec3i = Vector3[int]
